### ISSUE TITLE
Unsafe Integers

### ISSUE THREAD

**Issue by DavidVorick** [NebulousLabs/Sia#29]
Right now all of the integers that we use are unsafe. Meaning they'll overflow and golang won't warn us or provide an error.

So you could have an input of 50, and an output of uint64 max and a different output of 51, and the program will think that the total is 50 and will accept the transaction.

so... we need to add checks in that provent integer overflows in all situations. Maybe the 'Currency' type can be extended to return an error if there's an overflow.

**Comment by DavidVorick** [comment]
This is the next thing that needs to happen. When we're done with our current PR's we should attack this.

**Comment by lukechampine** [comment]
I'm starting to implement 128 bit integers. Currently they look like:
```go
// An int128 is a 128-bit signed integer. Operations on int128s are performed
// via math/big.
//
// The int128 also keeps track of whether an overflow has occurred during
// arithmetic operations. Once the 'overflow' flag has been set to true, it
// can never be reset; a new int128 must be created. Callers can check for
// overflow using the Overflow method. This allows arithmetic operations to be
// chained together without needing to check an error value after each
// operation.
type int128 struct {
	b  [16]byte
	of bool // has an overflow ever occurred?
}
```
I like the use of an overflow flag instead, since you only have to check for overflow once. As an optimization, we could even make `Add`, `Sub`, etc. no-ops if an overflow has previously occurred. There is one potential downside, which is that you have to explicitly check for an overflow rather than having an `err` shoved in your face.

Another problem is with the type system. Since `Currency` is now defined as a `int128`, you don't get the `Add`, `Sub`, etc. methods for free. You'd have to define `Currency` as a struct with an embedded `int128`. This goes for any other types that will be using `int128` as well. This would be a non-issue if `Currency` is the only 128-bit type, though.

Finally, there's the issue of pointers. The `math/big` library works almost exclusively with pointers. We could do the same, but it complicates things somewhat; do we define `Currency` as a `*int128`, or do we just use `*Currency`?

**Comment by DavidVorick** [comment]
Currency is the only 128bit integer that we use, so you can just call it a Currency.

I don't like using an overflow bool. Because that means you have to explicitly check, which means you might forget to check. And might forget is an unfortunate circumstance.

I know it means a lot more error checking lines, but I think it's the safer and smarter option.

**Comment by lukechampine** [comment]
Another approach would be to just use `big.Int` everywhere to do math, and then have a `BigToCurrency` function that converts to the `Currency` type and returns an error on overflow. That might make more sense than writing a bunch of pass-through functions anyway.

No way am I making functions like `Add` return an error though.

**Comment by DavidVorick** [comment]
fixed by #314
